src/app/app-platform/ContentAppPlatform.cpp

```diff
@@ -387,24 +388,106 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
+    mBindingEndpointId      = bindingEndpointId;
+    mBindingClusterId       = bindingClusterId;
+    mBindingSucessCallback  = successCb;
+    mBindingFailureCallback = failureCb;
+
+    CHIP_ERROR error = ReadFabricsList(device);
+    if (error != CHIP_NO_ERROR)
+    {
+        mBindingDevice = nullptr;
+    }
+    ChipLogDetail(Controller, "CreateBindingWithCallback: Sent read fabrics list request, waiting for response");
+    return error;
+}
 
-    Binding::Commands::Bind::Type request;
-    request.nodeId     = bindingNodeId;
-    request.groupId    = bindingGroupId;
-    request.endpointId = bindingEndpointId;
-    request.clusterId  = bindingClusterId;
-    ReturnErrorOnFailure(cluster.InvokeCommand(request, this, successCb, failureCb));
+CHIP_ERROR ContentAppPlatform::ReadFabricsList(OperationalDeviceProxy * device)
+{
+    chip::Controller::OperationalCredentialsCluster cluster;
+    cluster.Associate(device, 0);
+    return cluster.ReadAttribute<FabricsListDecodableType>(this, OperationalCredentials::Id,
+                                                           OperationalCredentials::Attributes::FabricsList::Id,
+                                                           OnReadFabricsListSuccess, OnReadFabricsListFailure);
+}
 
-    ChipLogDetail(Controller, "CreateBindingWithCallback: Sent Bind command request, waiting for response");
-    return CHIP_NO_ERROR;
+void ContentAppPlatform::OnReadFabricsListSuccess(void * context, const FabricsListDecodableType & data)
+{
+    ContentAppPlatform * obj = static_cast<ContentAppPlatform *>(context);
+    auto iter                = data.begin();
+    FabricInfo * fabric =
+        Server::GetInstance().GetFabricTable().FindFabricWithCompressedId(obj->mBindingDevice->GetPeerId().GetCompressedFabricId());
+
+    if (fabric == nullptr)
+    {
+        obj->NotifyBindingFailed(CHIP_ERROR_NOT_FOUND);
+    }
+
+    while (iter.Next())
+    {
+        PeerId peer;
+        fabric->GetCompressedId(iter.GetValue().fabricId, iter.GetValue().nodeId, &peer);
+        if (peer == obj->mBindingDevice->GetPeerId())
+        {
+            obj->SendBindingWriteRequest(iter.GetValue().fabricIndex);
+            return;
+        }
+    }
+    obj->NotifyBindingFailed(CHIP_ERROR_NOT_FOUND);
+}
+
+void ContentAppPlatform::OnReadFabricsListFailure(void * context, CHIP_ERROR error)
+{
+    ContentAppPlatform * obj = static_cast<ContentAppPlatform *>(context);
+    obj->NotifyBindingFailed(error);
+}
+
+void ContentAppPlatform::NotifyBindingFailed(CHIP_ERROR error)
+{
+    mBindingDevice = nullptr;
+    mBindingFailureCallback(nullptr, error);
+}
+
+void ContentAppPlatform::SendBindingWriteRequest(FabricIndex fabricIndex)
+{
+    chip::Controller::BindingCluster cluster;
+    cluster.Associate(mBindingDevice, mDeviceEndpointId);
+    Binding::Structs::TargetStruct::Type entries[1];
+
+    if (mBindingNodeId != kUndefinedNodeId)
+    {
+        entries[0] = Binding::Structs::TargetStruct::Type{
```

Comment:
I would vastly prefer it if we set the members of `entries[0]` by name here, so it's clear that we are not mixing things up (and e.g. putting the endpoint id where the group id should go, because those happen to be the same type).

Note that all the optionals are already initialized to NullOptional, so we'd just need to Emplace() the things that have values.